Draft: Feature/token search

tukantje opened the pull request:
# Summary

<<if there's an issue>>Fixes #issueNumber

*High-level description of what your changes are accomplishing*

*Add screenshots if applicable. Images are nice :)*

  # To Test

1. <<Step one>> Open the page `about`
- [ ] <<What to expect?>> Verify it contains about information...
- [ ] Checkbox Style list of things a QA person could verify, i.e.
- [ ] Should display Text Input our storybook
- [ ] Input should not accept Numbers
2. <<Step two>> ...

  # Background

*Optional: Give background information for changes you've made, that might be difficult to explain via comments*

shoom3301 commented on src/custom/components/SearchModal/CurrencySearch/useTokenSearch.ts:
```diff
@@ -0,0 +1,74 @@
+  })
+)
+
+export function useTokenSearch(_query: string, existingTokens: Map<string, boolean>) {
```
Please, set the return type

tukantje replied:
Thank you - added it.

alfetopito commented:
## Comment about the PR

Very nice! There are test steps and a description 👍 

As for the linked issue, `Adds` is not a special word Github recognizes. Try `Closes` for example. See here the full reference https://docs.github.com/en/get-started/writing-on-github/working-with-advanced-formatting/using-keywords-in-issues-and-pull-requests

## Comment about the behaviour

Working great!
The thing that I find missing however is the token image.
In this example, I can see the response has an image but it's not loaded. Any idea why?
![Screenshot 2023-03-09 at 09 25 54](https://user-images.githubusercontent.com/43217/223979476-05748e38-366c-442d-84a7-b18cf216cfd4.png)

## Comments about the code

None yet, will review it next

tukantje commented:
> ## Comment about the PR
> 
> Very nice! There are test steps and a description 👍
> 
> As for the linked issue, `Adds` is not a special word Github recognizes. Try `Closes` for example. See here the full reference https://docs.github.com/en/get-started/writing-on-github/working-with-advanced-formatting/using-keywords-in-issues-and-pull-requests
> ## Comment about the behaviour
> 
> Working great! The thing that I find missing however is the token image. In this example, I can see the response has an image but it's not loaded. Any idea why? ![Screenshot 2023-03-09 at 09 25 54](https://user-images.githubusercontent.com/43217/223979476-05748e38-366c-442d-84a7-b18cf216cfd4.png)
> ## Comments about the code
> 
> None yet, will review it next

Changed the wording in the PR.

Good one - I see some images being loaded but not all; looking into it. Image does exist, so most likely I missed something there.

alfetopito commented on src/custom/components/SearchModal/CurrencySearch/useTokenSearch.ts:
```diff
@@ -0,0 +1,74 @@
+}
+
+async function fetchTokens(query: string): Promise<FetchTokensApiResult> {
+  const response = await fetch(`${BASE_URL}${query}`)
+  const data = await response.json()
+
+  return data
```
What happens if the request fails (5xx, 4xx) or the response does not contain a json?

Also, no need for the intermediate `data` variable, can return `await response.json()` directly.

tukantje replied:
Loadable essentially wraps this for us, where they return state with `hasError` instead of `hasData`.

Currently I am treating it as if we can't get it, just stay silent (as this is just an external source) - which seems to be the most graceful way of failing there.

However if we'd like to we could rely on the state variable from the loadable atom to craft anything pretty much.

> Also, no need for the intermediate data variable, can return await response.json() directly.
Replacing the data variable, great point.

alfetopito commented on src/custom/components/SearchModal/CurrencySearch/useTokenSearch.ts:
```diff
@@ -0,0 +1,74 @@
+
+const searchQuery = atom('')
+
+const tokensData = loadable(
```
Async atom, nice!

tukantje replied:
>  So, I change the request to be in a follow up PR rather than here :)
Great point, happy to follow up on this. I see quite a bit of refactor potential here as well; so can follow up with that (and should help reduce our dependency on the uniswap's base)

alfetopito commented on src/custom/components/SearchModal/CurrencySearch/useTokenSearch.ts:
```diff
@@ -0,0 +1,74 @@
+  const [query, setQuery] = useAtom(searchQuery)
+  // query is being used through jotai
+  // eslint-disable-next-line react-hooks/exhaustive-deps
+  const [result] = useAtom(useMemo(() => tokensData, [query]))
```
Nitpick: if only reading, you can use `useAtomValue` instead

tukantje replied:
Great point - updated

alfetopito reviewed (commented):
Looks very good and clean, loved it!

One think that I'm going to ask is to move those files out of `custom` subfolder into `cow-react`.

As you are aware, we broke out with Uniswap's original structure, so we should "own" our code.
That means turning the `Mod` components into our own version of them.
Which is essentially the same code we have minus the duplication and the overrides.

Still, I also feel this is out of the scope of this PR which is very nice and easy to understand. Doing what I suggested above here would just add noise.
So, I change the request to be in a follow up PR rather than here :)

elena-zh commented:
Hey @tukantje , thank you for the great PR description!
I have done some tests, and it works great! However, some issues I have faced: 
1. The app is crashed when I enter a space into the search field  :)
![search by space](https://user-images.githubusercontent.com/70885163/224059495-91b5257c-460a-4034-a0c5-24dc9d902abe.png)
2. Search by a long token name is not working  :( (I reported it in #1915 )
![no token name](https://user-images.githubusercontent.com/70885163/224059766-9f0ff658-11d6-4523-a653-bc53118c061d.png)
But I still can find it by the address. Question: maybe we should add a section's header 'Additional Results from External Sources' when we search by address, and a token is not in any list?
![address-external](https://user-images.githubusercontent.com/70885163/224060239-43f52e29-d82e-40f6-9385-b4ae878150fb.png)
3. When importing a token from an external souse, maybe we should change the Import token modal description, WDYT?
![import token](https://user-images.githubusercontent.com/70885163/224060392-43744888-405c-4062-9595-c32de8a76c01.png)
4. Should we display a search result with the list of tokens that is not related to the currently selected network? 
![image](https://user-images.githubusercontent.com/70885163/224061313-ecebb5b8-dc2f-4d62-b119-afebc98c2142.png)
![image](https://user-images.githubusercontent.com/70885163/224061387-7d951597-8b03-4b85-9aeb-34d30b4b919a.png)
I'd show the message like this instead:
![image](https://user-images.githubusercontent.com/70885163/224061656-951bb60f-50f4-4a8f-af58-2129dcf372af.png)

And some nice issues that would be great to take a look at:
 - https://github.com/cowprotocol/cowswap/issues/23
 - https://github.com/cowprotocol/cowswap/issues/163


Thanks!

tukantje commented:
> Hey @tukantje , thank you for the great PR description! I have done some tests, and it works great! However, some issues I have faced:
> 
>     1. The app is crashed when I enter a space into the search field  :)
>        ![search by space](https://user-images.githubusercontent.com/70885163/224059495-91b5257c-460a-4034-a0c5-24dc9d902abe.png)

I have added an additional check for this, however I cannot replicate it unfortunately. 

>     2. Search by a long token name is not working  :( (I reported it in [Add possibility to search for a token by a long name #1915](https://github.com/cowprotocol/cowswap/issues/1915) )
>        ![no token name](https://user-images.githubusercontent.com/70885163/224059766-9f0ff658-11d6-4523-a653-bc53118c061d.png)

Changes I've made here are about adding UniSwap as an external source of tokens. UniSwap itself doesn't seem to be returning anything for this token;
<img width="536" alt="Screenshot 2023-03-09 at 22 22 31" src="https://user-images.githubusercontent.com/5172699/224163056-49689e2b-ec70-417d-8aaf-f383e3b60934.png">

Regarding the rest of the code not being able to find, I'd be happy to follow up on #1915 

>        But I still can find it by the address. Question: maybe we should add a section's header 'Additional Results from External Sources' when we search by address, and a token is not in any list?
>        ![address-external](https://user-images.githubusercontent.com/70885163/224060239-43f52e29-d82e-40f6-9385-b4ae878150fb.png)

What would you imagine the section to be called? For background, once we are doing an address search, whole section works differently (i.e. we don't check external sources)

>     3. When importing a token from an external souse, maybe we should change the Import token modal description, WDYT?
>        ![import token](https://user-images.githubusercontent.com/70885163/224060392-43744888-405c-4062-9595-c32de8a76c01.png)

I am not sure what the process there is, however personally I don't really see too much of a benefit there. I think the important distinction is the token not being on an active list (i.e. you are importing something, are you sure of what you are doing?)

I think the same situation applies to external sources, as well as inactive lists but I might be missing something?

>     4. Should we display a search result with the list of tokens that is not related to the currently selected network?
>        ![image](https://user-images.githubusercontent.com/70885163/224061313-ecebb5b8-dc2f-4d62-b119-afebc98c2142.png)
>        ![image](https://user-images.githubusercontent.com/70885163/224061387-7d951597-8b03-4b85-9aeb-34d30b4b919a.png)
>        I'd show the message like this instead:
>        ![image](https://user-images.githubusercontent.com/70885163/224061656-951bb60f-50f4-4a8f-af58-2129dcf372af.png)
> 

Great point. Added a filter for this.
<img width="488" alt="Screenshot 2023-03-09 at 22 35 33" src="https://user-images.githubusercontent.com/5172699/224165351-d236a440-c125-4486-8723-858da259c3ba.png">


> And some nice issues that would be great to take a look at:
> 
>     * [Token search address - Case sensitive #23](https://github.com/cowprotocol/cowswap/issues/23)
> 
>     * [Token list: token is not found when there are preceding/trailing spaces in the address #163](https://github.com/cowprotocol/cowswap/issues/163)
> 
> 
> Thanks!

Thank you! I'll follow up on these - however, changes here are mainly regarding adding an extra source, so I think it makes more sense if I follow up separately with these issues.

elena-zh commented:
Hey @tukantje , 
the 1st and the 4th cases are fixed, thank you!

> What would you imagine the section to be called? For background, once we are doing an address search, whole section works differently (i.e. we don't check external sources)

I mean this name of the section: "Expanded results from inactive Token Lists"

> I think the same situation applies to external sources, as well as inactive lists but I might be missing something?

My point here is: currently, we search for tokens from active token lists first, and then from 'inactive'. And test on the import modal looks logical now to me as we import tokens from **inactive** lists.

But in this PR we are adding a new source - 'External'. So I'm proposing here to separate these cases and change the text on the import modal when we import a token from an external source. 
Anyways, low priority. Let me know please if I need to create a separate issue for this. 

Thanks!

tukantje commented:
@elena-zh I think it would make sense if we can split these, based on priority, I have still some follow up work here for sure, so can address them separately.

elena-zh reviewed (approved):
I have created #2153 and #2152 enhancements' tasks for the cases I proposed above.

tukantje commented on src/custom/components/SearchModal/CurrencySearch/useTokenSearch.ts:
```diff
@@ -0,0 +1,75 @@
+      .map(({ chainId, address, decimals, symbol, name }) => new Token(chainId, address, decimals, symbol, name))
+  }, [result, existingTokens, currentChainId])
+
+  useEffect(() => {
+    setQuery(_query)
+  }, [setQuery, _query])
```
Great point!

Resolving this as we've discussed on Slack - the value here is supplied already debounced, so should not cause trouble.

anxolin commented on src/custom/components/SearchModal/CurrencySearch/useTokenSearch.ts:
```diff
@@ -0,0 +1,75 @@
+  const response = await fetch(`${BASE_URL}${query}`)
+
+  return await response.json()
+}
```
In principle,  all of the above lines could be isolated in the API layer.

Most of the new code should go to `cow-react`.

<img width="577" alt="image" src="https://user-images.githubusercontent.com/2352112/227593302-871b1d57-56c0-4b33-a212-e16de5e06cc8.png">

tukantje replied:
Good point - I'll follow up on this to move it there.

anxolin commented on src/custom/components/SearchModal/CurrencySearch/useTokenSearch.ts:
```diff
@@ -0,0 +1,75 @@
+
+    return undefined
+  })
+)
```
This is state. Normally we place these ones in a module's state.

Just curious, why we use joitai state here? 
what happen if we use this hook in two different parts of the UI? that useEffect will trigger the setQuery and they will be fighting for the right to update the atom?

anxolin reviewed (commented):
I might be wrong, but It feels a bit overcomplicated. 
I don't see the point of using joitai state here.
